xapian-letor/api/featurelist_internal.cc

```diff
@@ -0,0 +1,181 @@
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301
+ * USA
+ */
+
+#include <config.h>
+
+#include "xapian-letor/featurelist.h"
+#include "featurelist_internal.h"
+
+#include <cmath>
+#include <cstdio>
+#include <cstdlib>
+#include <cstring>
+#include "debuglog.h"
+
+using namespace std;
+using namespace Xapian;
+
+void
+FeatureList::Internal::set_database(const Xapian::Database & db)
+{
+    featurelist_db = db;
+}
+
+void
+FeatureList::Internal::set_query(const Xapian::Query & query)
+{
+    featurelist_query = query;
+}
+
+void
+FeatureList::Internal::set_doc(const Xapian::Document & doc)
+{
+    featurelist_doc = doc;
+}
+
+void
+FeatureList::Internal::compute_statistics(const Xapian::Query & letor_query,
+					  const Xapian::Database & letor_db,
+					  const Xapian::Document & letor_doc)
+{
+    set_query(letor_query);
+    set_doc(letor_doc);
+    set_database(letor_db);
+
+    compute_termfreq();
+    compute_inverse_doc_freq();
+    compute_doc_length();
+    compute_collection_length();
+    compute_collection_termfreq();
+}
+
+void
+FeatureList::Internal::compute_termfreq()
+{
+    std::map<std::string, Xapian::termcount> tf;
+
+    Xapian::TermIterator docterms = featurelist_doc.termlist_begin();
+    for (Xapian::TermIterator qt = featurelist_query.get_unique_terms_begin();
+	 qt != featurelist_query.get_terms_end(); ++qt) {
+	docterms.skip_to(*qt);
+	if (docterms != featurelist_doc.termlist_end() && *qt == *docterms)
+	    tf[*qt] = docterms.get_wdf();
+    }
+    std::swap(termfreq, tf);
+}
+
+void
+FeatureList::Internal::compute_inverse_doc_freq()
+{
+    std::map<std::string, double> idf;
+
+    for (Xapian::TermIterator qt = featurelist_query.get_unique_terms_begin();
+	 qt != featurelist_query.get_terms_end(); ++qt) {
+	Xapian::doccount totaldocs = featurelist_db.get_doccount();
+	Xapian::doccount df = featurelist_db.get_termfreq(*qt);
+	if (df != 0)
+	    idf[*qt] = log10((double)totaldocs / (double)(1 + df));
+    }
+    std::swap(inverse_doc_freq, idf);
+}
+
+void
+FeatureList::Internal::compute_doc_length()
+{
+    std::map<std::string, Xapian::termcount> len;
+
+    Xapian::termcount title_len = 0;
+    Xapian::termcount whole_len = 0;
+    Xapian::TermIterator dt = featurelist_doc.termlist_begin();
+    // reach the iterator to the start of the title terms i.e. prefix "S"
+    dt.skip_to("S");
+    for ( ; dt != featurelist_doc.termlist_end(); ++dt) {
+	if ((*dt)[0] != 'S') {
+	    // We've reached the end of the S-prefixed terms.
+	    break;
+	}
+	title_len += dt.get_wdf();
+    }
+    len["title"] = title_len;
+    whole_len = featurelist_db.get_doclength(featurelist_doc.get_docid());
```

Comment:
It's better C++ style to declare a variable where you can initialise it, which you can for `whole_len`.

As the code is, someone might add or move some code which uses `whole_len` early in the function, and it'll still compile and run OK (but give the wrong results because it'll use `0` instead of the correct value, which hopefully the testsuite or the developer will notice, but it could easily result in a bug exposed to users).

If you instead declare and initialise `whole_len` here, then any attempt to use it before it's set to the correct value results in a compile-time error.

This same issue is repeated below - please find and address all cases.